alica_engine/include/engine/blackboard/Blackboard.h

```diff
@@ -23,25 +93,143 @@ class BlackboardImpl
     template <typename T>
     const T& get(const std::string& key) const
     {
-        return std::any_cast<const T&>(vals.at(key));
+        try {
+            return convert<T>::get(vals.at(key));
+        } catch (const std::bad_variant_access& e) {
+            throw BlackboardTypeMismatch(getNameFromType<T>(), getBlackboardValueType(key));
+        } catch (const std::bad_any_cast& e) {
+            throw BlackboardTypeMismatch(getNameFromType<T>(), getBlackboardValueType(key));
+        }
     }
     template <typename T>
-    T& get(const std::string& key)
+    const T& get(const std::string& key)
     {
-        return std::any_cast<T&>(vals.at(key));
+        try {
+            return convert<T>::get(vals.at(key));
+        } catch (const std::bad_variant_access& e) {
+            throw BlackboardTypeMismatch(getNameFromType<T>(), getBlackboardValueType(key));
+        } catch (const std::bad_any_cast& e) {
+            throw BlackboardTypeMismatch(getNameFromType<T>(), getBlackboardValueType(key));
+        }
     }
-    std::any& get(const std::string& key) { return vals.at(key); }
-    const std::any& get(const std::string& key) const { return vals.at(key); }
+    BlackboardValueType& get(const std::string& key) { return vals.at(key); }
+    const BlackboardValueType& get(const std::string& key) const { return vals.at(key); }
 
-    void set(const std::string& key, const std::any& value) { vals.at(key) = value; }
+    template <class T>
+    void set(const std::string& key, const T& value)
+    {
+        if (getBlackboardValueNode(key).Type() == YAML::NodeType::Null) {
+            // value is not predefined in pml file
+            if (getNameFromType<T>() == "std::any") {
+                // if type is std::any, wrap into std::any
+                vals.emplace(key, std::any{value});
+            } else {
+                vals.emplace(key, value);
+            }
+        } else {
+            // value is predefined in pml file
+            std::string inputType = getNameFromType<T>();
+            std::string pmlType = getBlackboardValueType(key);
+            if (pmlType == "std::any") {
+                // insert as std::any with type T
+                vals.at(key) = std::any{value};
+            } else if (pmlType == inputType) {
+                vals.at(key) = value;
+            } else {
+                // types dont match, throw exception
+                throw BlackboardTypeMismatch(inputType, pmlType);
+            }
+        }
+    }
 
     bool hasValue(const std::string& key) const { return vals.count(key); }
     void removeValue(const std::string& key) { vals.erase(key); }
 
+    void initDefaultValues();
+
     void clear() { vals.clear(); }
     bool empty() const { return vals.empty(); }
     size_t size() const { return vals.size(); }
-    std::unordered_map<std::string, std::any> vals;
+    std::unordered_map<std::string, BlackboardValueType> vals;
+    YAML::Node node;
+
+private:
+    friend BlackboardUtil;
```

Comment:
I think we really have a need to separate the blackboard from implementation at this point.  Ideally implementation with separate header file, that is included by `Blackboard.cpp` and not `Blackboard.h` (maybe pimpl?)

Reply:
most of the functions are templates but yeah, can look into how to do this. But can we defer to a separate PR?